From 978e5532c1c6b6f248ec7b62cd053c4a9e1dd10f Mon Sep 17 00:00:00 2001
From: emaste <emaste@FreeBSD.org>
Date: Mon, 17 Apr 2017 13:22:39 +0000
Subject: [PATCH 8/9] bsdgrep: for -r, use the working directory if none
 specified

This is more sensible than the previous behaviour of grepping stdin,
and matches newer GNU grep behaviour.

PR:		216307
Submitted by:	Kyle Evans <kevans91 at ksu.edu>
Reviewed by:	cem, emaste, ngie
Relnotes:	Yes
Differential Revision:	https://reviews.freebsd.org/

Change-Id: I9dc42c4105c35c42c6b195c880c70f4398852886
---
 toolbox/upstream-netbsd/usr.bin/grep/grep.c | 2 +-
 toolbox/upstream-netbsd/usr.bin/grep/util.c | 5 ++++-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/toolbox/upstream-netbsd/usr.bin/grep/grep.c b/toolbox/upstream-netbsd/usr.bin/grep/grep.c
index 1ea6ed3aa..ae0f99af9 100644
--- a/toolbox/upstream-netbsd/usr.bin/grep/grep.c
+++ b/toolbox/upstream-netbsd/usr.bin/grep/grep.c
@@ -685,7 +685,7 @@ main(int argc, char *argv[])
 	if ((aargc == 0 || aargc == 1) && !Hflag)
 		hflag = true;
 
-	if (aargc == 0)
+	if (aargc == 0 && dirbehave != DIR_RECURSE)
 		exit(!procfile("-"));
 
 	if (dirbehave == DIR_RECURSE)
diff --git a/toolbox/upstream-netbsd/usr.bin/grep/util.c b/toolbox/upstream-netbsd/usr.bin/grep/util.c
index ecd948da4..c74d99f74 100644
--- a/toolbox/upstream-netbsd/usr.bin/grep/util.c
+++ b/toolbox/upstream-netbsd/usr.bin/grep/util.c
@@ -117,6 +117,7 @@ grep_tree(char **argv)
 	char *d, *dir = NULL;
 	int c, fts_flags;
 	bool ok;
+	const char *wd[] = { ".", NULL };
 
 	c = fts_flags = 0;
 
@@ -134,7 +135,9 @@ grep_tree(char **argv)
 
 	fts_flags |= FTS_NOSTAT | FTS_NOCHDIR;
 
-	if (!(fts = fts_open(argv, fts_flags, NULL)))
+	fts = fts_open((argv[0] == NULL) ?
+	    ((char * const *)(uintptr_t)(const void *)wd) : argv, fts_flags, NULL);
+	if (fts == NULL)
 		err(2, "fts_open");
 	while ((p = fts_read(fts)) != NULL) {
 		switch (p->fts_info) {
-- 
2.17.1

